File: TypeScript/src/workTrackerMcpAgent.ts
import { randomUUID } from 'crypto';
import * as readline from 'readline/promises';

import { BaseMessage, HumanMessage, SystemMessage, ToolMessage, AIMessageChunk } from "@langchain/core/messages";
import { Annotation, END, MemorySaver, messagesStateReducer, START, StateGraph } from "@langchain/langgraph";
import { initModelProvider, LLMProvider } from "./llmProviders";
import { BaseChatModel } from "@langchain/core/language_models/chat_models";

import { Runnable } from "@langchain/core/runnables";
import { StructuredTool, tool } from "@langchain/core/tools";

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

import { getAllLocalAgentTools } from "./agentTools.js";
import dedent from 'dedent';


const DEFAULT_CHAT_MODEL_TEMPERATURE = 0.0;

const WorkTrackerMCPAgentStateAnnotation = Annotation.Root({
    // messagesStateReducer acts exactly like `add_messages`
    messages: Annotation<BaseMessage[]>({
        reducer: messagesStateReducer,
        default: () => [],
    }),
    loopCount: Annotation<number>({
        reducer: (oldState, newState) => newState, // Overwrites (default behavior)
        default: () => 0,
    })
});

type WorkTrackerMCPAgentState = typeof WorkTrackerMCPAgentStateAnnotation.State;


export class WorkTrackerMCPAgent {
    public llmProvider: LLMProvider;
    public mcpServerUrl: string;
    public chatModel: string;
    public chatModelTemperature: number;

    private llm: BaseChatModel;
    private llmWithTools!: Runnable<any, any>;
    private graph!: ReturnType<WorkTrackerMCPAgent["buildGraph"]>;
    private memoryCheckpointer: MemorySaver;

    private mcpClient!: Client;

    private discoveredTools!: StructuredTool[];
    private localTools!: StructuredTool[];
    private allTools!: StructuredTool[];    
    private toolsRegistry!: Record<string, StructuredTool>;

    private static readonly MAX_SIMULTANEOUS_TOOL_CALLS = 5;
    private static readonly MAX_TOOL_CALLS_ROUNDS = 10;

    /**
     * The private constructor of the WorkTrackerMCPAgent class.
     * @param llmProvider 
     * @param chatModel 
     * @param mcpServerUrl 
     * @param chatModelTemperature 
     */
    private constructor(
        llmProvider: LLMProvider, 
        chatModel: string, 
        mcpServerUrl: string, 
        chatModelTemperature: number = DEFAULT_CHAT_MODEL_TEMPERATURE
    ) {
        this.chatModel = chatModel;
        this.chatModelTemperature = chatModelTemperature;
        this.llmProvider = llmProvider;
        this.mcpServerUrl = mcpServerUrl;

        this.llm = initModelProvider(this.llmProvider, this.chatModel, this.chatModelTemperature);

        this.memoryCheckpointer = new MemorySaver();
    }
    /**
     * The public factory of the WorkTrackerMCPAgent class.
     * @param llmProvider 
     * @param chatModel 
     * @param mcpServerUrl 
     * @param chatModelTemperature 
     */
    public static async create(
        llmProvider: LLMProvider, 
        chatModel: string, 
        mcpServerUrl: string, 
        chatModelTemperature: number = DEFAULT_CHAT_MODEL_TEMPERATURE
    ): Promise<WorkTrackerMCPAgent> {
        const instance = new WorkTrackerMCPAgent(llmProvider, chatModel, mcpServerUrl, chatModelTemperature);
        await instance.initMCPAndGraph();
        return instance;
    }

    /**
     * Process a single user query.
     * @param userMessage The user query.
     * @param threadId Optional ID of the chat thread to continue on.
     * @returns Promise with the resulting Thread ID, which can be later used to resume the in-memory saved chat thread.
     */
    public async runSingleQuery(userMessage: string, threadId: string | undefined = undefined): Promise<string> {
        threadId ??= randomUUID();
        const config = { configurable: { thread_id: threadId } };

        const initialState: WorkTrackerMCPAgentState = {
            messages: [new HumanMessage(userMessage)],
            loopCount: 0,
        }

        const result = await this.graph.invoke(initialState, config) as WorkTrackerMCPAgentState;

        console.log(`\nAssistant: ${result.messages}`)

        return threadId;
    }

    /**
     * Run a CLI streaming user chat loop.
     * Exit from the loop by pressing Ctrl+C during the agent's thinking or answering,
     * or by pressing Ctrl+D when waiting for a user input.
     * @param cancellationCallback The callback to be called to check if a termination has been requested.
     * @param threadId Optional ID of the chat thread to continue on.
     * @returns Promise with the resulting Thread ID, which can be later used to resume the in-memory saved chat thread.
     */
    public async runStreamingLoop(cancellationCallback: () => boolean, threadId: string | undefined = undefined): Promise<string> {
        threadId ??= randomUUID();
        const config = { configurable: { thread_id: threadId } };

        // Print the initial info to the user
        console.log(dedent(`
            --- Session Started ---
        
            In order to terminate the agent,
            press Ctrl+C when the agent is thinking or answering
            or Ctrl+D when waiting for the user input.
        `).trim() + "\n")

        // Create the readline interface
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });

        try {
            while (!cancellationCallback()) {
                // Await the user input
                const userMessage = await rl.question("User:      ");

                // Handle Ctrl+D
                // (EOF in Node.js readline often returns an empty string or requires a specific event listener, 
                // but we can also support an explicit exit command)
                if (userMessage.trim().toLowerCase() === 'exit') {
                    break;
                }

                const passedState = {
                    messages: [new HumanMessage(userMessage)],
                    loopCount: 0
                };

                process.stdout.write("\nThinking...\r"); // So that it can be rewritten

                const stream = await this.graph.stream(
                    passedState,
                    { ...config, streamMode: "messages" }
                );

                let prefixPrinted = false;

                for await (const chunk of stream) {
                    if (cancellationCallback()) {
                        break;
                    }

                    const [msg, metadata] = chunk;

                    // We target content chunks originating specifically from our 'agent' node
                    if (metadata.langgraph_node === WorkTrackerMCPAgent.AGENT_NODE_NAME) {
                        // Check if the chunk contains text content (ignores empty tool-call chunks)
                        if (msg.content && typeof msg.content === "string") {
                            if (!prefixPrinted) {
                                // Clear the "Thinking..." line
                                process.stdout.write("\r\x1b[K"); 
                                process.stdout.write("Assistant: ");
                                prefixPrinted = true;
                            }

                            process.stdout.write(msg.content);
                        }
                    }
                }

                if (prefixPrinted) {
                    console.log("\n\n");
                }
            }
        } finally {
            // ALWAYS close the interface in a finally block to prevent memory leaks or hanging processes
            rl.close();
        }

        return threadId;
    }

    // --- MCP Discovery and Graph initialization ---

    /**
     * Asynchronously connects to the Work Tracker MCP server, discovers tools, and builds the graph.
     */
    private async initMCPAndGraph() : Promise<void> {
        // Connect to the MCP server
        console.log(`Connecting to Work Tracker MCP server at '${this.mcpServerUrl}'...`);
        const transport = new StreamableHTTPClientTransport(new URL(this.mcpServerUrl));

        this.mcpClient = new Client(
            { name: "work-tracker-mcp-agent", version: "1.0.0" },
            { capabilities: { } } // The client doesn't need to offer any special capabilities
        );

        await this.mcpClient.connect(transport);

        // Automatically fetch all tools exposed by the MCP server (list_spaces, start_work_day, etc.)
        const mcpToolsResponse = await this.mcpClient.listTools();

        // Translate raw MCP tools into LangChain-compatible tools
        this.discoveredTools = mcpToolsResponse.tools.map(mcpTool => {
            return tool(
                async (args: Record<string, unknown>) => {
                    // When LangGraph calls this tool, we forward the call to the Dart server
                    const result = await this.mcpClient.callTool({
                        name: mcpTool.name,
                        arguments: args
                    });

                    const contentBlocks = result.content as Array<{ type: string; text?: string }>;
                    
                    // MCP returns an array of content blocks; we extract the text
                    // Check if the result has content, and map text properties
                    return contentBlocks
                        .filter(c => c.type === "text")
                        .map(c => c.text)
                        .join("\n");
                },
                {
                    name: mcpTool.name,
                    description: mcpTool.description || `Executes the ${mcpTool.name} action via MCP.`,
                    // LangChain.js allows us to pass the raw JSON Schema
                    // directly from the MCP server without having to parse it into Zod!
                    schema: mcpTool.inputSchema as any, 
                }
            );
        });

        // Combine local tools and remote MCP tools
        this.localTools = getAllLocalAgentTools();
        this.allTools = [...this.discoveredTools, ...this.localTools];

        // Build the tools registry and and bind them to the LLM
        this.toolsRegistry = {};
        for (const t of this.allTools) {
            this.toolsRegistry[t.name] = t;
        }

        this.llmWithTools = this.llm.bindTools!(Object.values(this.toolsRegistry));

        // Now that tools are bound, compile the graph
        this.graph = this.buildGraph();

        console.log(`Successfully connected! Discovered ${this.discoveredTools.length} MCP tools and adding ${this.localTools.length} local tools.`);
    }

    // --- The nodes ---
    private static readonly AGENT_NODE_NAME = "agent" as const;
    private static readonly TOOLS_NODE_NAME = "tools" as const;

    /**
     * Node: Invokes the LLM with the current conversation history, prepending the system message.
     * @param state The input graph state.
     * @returns The resulting graph state partial update.
     */
    private async agentNode(state: WorkTrackerMCPAgentState): Promise<Partial<WorkTrackerMCPAgentState>> {
        const localToolsNames = [];
        for (const t of this.localTools) {
            localToolsNames.push(t.name);
        }
        const toolNamesString = this.localTools.map(tool => tool.name).join("`, `");
        const localToolsAddend = this.localTools.length > 0 ? ` along with a set of local tools (\`${toolNamesString}\`)` : "";

        const systemMessage = dedent(`
            You are a highly capable, integrated executive assistant connected directly to the user's Work Tracker application via the Model Context Protocol (MCP).
            Your primary role is to seamlessly manage their workspaces, tasks, work days, and time logging.

            ### Operational Guidelines:
            1. **Auto-Discovery:** You have access to a suite of Work Tracker tools (e.g., \`list_actionable_tasks\`, \`start_work_day\`, \`start_task_work\`)${localToolsAddend}. 
                Use them proactively to fulfill the user's intent.
            2. **State Dependencies:** If the user asks to "start working on a task", you MUST check if there is an active work day using \`get_current_work_status\`.
                If there isn't one, invoke \`start_work_day\` before invoking \`start_task_work\`.
            3. **Data Retrieval:**
                If asked for a summary, utilize tools like \`get_time_report\` or \`list_work_days\` to provide accurate, up-to-date metrics. 
            4. **Destructive Actions:** Proceed carefully with deletions (\`delete_task\`, \`delete_work_day\`). 
                If the user request is ambiguous, ask for confirmation before executing.
            5. **Efficiency:** You may execute up to ${WorkTrackerMCPAgent.MAX_SIMULTANEOUS_TOOL_CALLS} tool calls simultaneously and before giving an answer to the user, in only at most ${WorkTrackerMCPAgent.MAX_TOOL_CALLS_ROUNDS} rounds. 
                Therefore if you were about to call tools in more than ${WorkTrackerMCPAgent.MAX_TOOL_CALLS_ROUNDS}, answer to the user after at most ${WorkTrackerMCPAgent.MAX_TOOL_CALLS_ROUNDS} tool call rounds and ask the user for a follow-up if they want to proceed.
                Do not invent task IDs; search for them if the user provides a name instead of an ID using \`search_tasks\`.

            Communicate concisely and professionally. You are assisting a power user who values speed and exactness.

            The current date and time is: ${new Date().toLocaleString(undefined, { 
                weekday: 'long', 
                year: 'numeric', 
                month: 'numeric', 
                day: 'numeric', 
                hour: 'numeric', 
                minute: 'numeric', 
                second: 'numeric' 
            })}.
        `).trim();

        const allMessages = [new SystemMessage(systemMessage), ...state.messages];

        // The LLM looks at the history and decides to either reply text or output a tool call.
        const response = await this.llmWithTools.invoke(allMessages);
        const currentLoops = state.loopCount;
        return {
            messages: [response],
            loopCount: currentLoops + 1,
        };
    }

    /**
     * Node: Manually loops through tool calls requested by the LLM.
     * @param state The input graph state.
     * @returns The resulting graph state partial update.
     */
    private async executeToolsNode(state: WorkTrackerMCPAgentState): Promise<Partial<WorkTrackerMCPAgentState>> {
        const lastMessage = state.messages[state.messages.length - 1];
        if (!(lastMessage instanceof AIMessageChunk && lastMessage.tool_calls && lastMessage.tool_calls.length > 0)) {
            return {};
        }

        // Programmatic Guardrail for max. parallelly queried items
        const lastToolCallsCount = lastMessage.tool_calls.length;
        if (lastToolCallsCount > WorkTrackerMCPAgent.MAX_SIMULTANEOUS_TOOL_CALLS) {
            const errorMessages: ToolMessage[] = [];
            for (const toolCall of lastMessage.tool_calls) {
                errorMessages.push(
                    new ToolMessage(
                        `Error: Maximum of ${WorkTrackerMCPAgent.MAX_SIMULTANEOUS_TOOL_CALLS} items allowed. You queried ${lastToolCallsCount}. Please self-correct by prioritizing the most important tool calls, combining queries if possible, or executing them in sequential batches instead. Do not ask the user for help.`,
                        toolCall.id!,
                    )
                );
            }
            return { messages: errorMessages };
        }

        // Setup lists to hold our async tasks and their corresponding IDs
        const tasks: Promise<any>[] = [];
        const toolCallIds: string[] = [];

        for (const toolCall of lastMessage.tool_calls) {
            const toolFunc = this.toolsRegistry[toolCall.name];
            if (toolFunc) {
                tasks.push(toolFunc.invoke(toolCall.args));
                toolCallIds.push(toolCall.id!);
            }
        }

        // Execute all tools simultaneously, waiting for all to finish
        // Promise.allSettled prevents one failing tool from crashing the whole batch
        const results = await Promise.allSettled(tasks);

        const toolResponses = results.map((result, index) => {
            const callId = toolCallIds[index];
            let content: string;

            if (result.status === "rejected") {
                const errorMessage = result.reason instanceof Error
                    ? result.reason.message
                    : String(result.reason);
                content = `Tool execution failed: ${errorMessage}`;
            }
            else {
                // LangChain tools generally return strings, but we stringify just in case
                content = typeof result.value === "string"
                    ? result.value
                    : JSON.stringify(result.value);
            }

            return new ToolMessage(content, callId);
        });

        return { messages: toolResponses };
    }

    /**
     * Conditional edge router: Custom router acting as a programmatic circuit breaker.
     * @param state The input graph state.
     * @returns The name of the node to be routed to, or the END.
     */
    private routeOrBreak(state: WorkTrackerMCPAgentState): string {
        // Circuit Breaker Triggered: Force exit if agent loops too many times
        if (state.loopCount > WorkTrackerMCPAgent.MAX_TOOL_CALLS_ROUNDS) {
            console.log(`⚠️ Circuit breaker triggered! Agent exceeded ${WorkTrackerMCPAgent.MAX_TOOL_CALLS_ROUNDS} loops.`);
            return END;
        }

        // Standard operational routing path
        const lastMessage = state.messages[state.messages.length - 1];
        if (lastMessage instanceof AIMessageChunk && lastMessage.tool_calls && lastMessage.tool_calls.length > 0) {
            return WorkTrackerMCPAgent.TOOLS_NODE_NAME;
        }

        return END;
    }

    // --- The graph ---

    /**
     * Builds the underlying LangGraph agent graph.
     */
    private buildGraph() {
        const builder = new StateGraph(WorkTrackerMCPAgentStateAnnotation)
            // Nodes
            .addNode(WorkTrackerMCPAgent.AGENT_NODE_NAME, this.agentNode.bind(this))
            .addNode(WorkTrackerMCPAgent.TOOLS_NODE_NAME, this.executeToolsNode.bind(this))

            // Edges - a chain with the tool call loop
            .addEdge(START, WorkTrackerMCPAgent.AGENT_NODE_NAME)
            .addConditionalEdges(WorkTrackerMCPAgent.AGENT_NODE_NAME, this.routeOrBreak.bind(this), {
                [WorkTrackerMCPAgent.TOOLS_NODE_NAME]: WorkTrackerMCPAgent.TOOLS_NODE_NAME,
                [END]: END,
            })
            .addEdge(WorkTrackerMCPAgent.TOOLS_NODE_NAME, WorkTrackerMCPAgent.AGENT_NODE_NAME);

        // Compile the graph
        return builder.compile({ checkpointer: this.memoryCheckpointer });
    }
}

File: Python/work_tracker_mcp_agent.py
from datetime import datetime
import uuid
import textwrap

import asyncio

from typing import Annotated, Any, Sequence, Optional, Callable, Self
from typing_extensions import TypedDict

from langgraph.graph import StateGraph, START, END
from langgraph.graph.state import CompiledStateGraph, Runnable
from langgraph.graph.message import add_messages
from langgraph.checkpoint.memory import InMemorySaver

from langchain_core.messages import AIMessage, SystemMessage, HumanMessage, BaseMessage, ToolMessage
from langchain_core.language_models.base import LanguageModelInput
from langchain_core.language_models.chat_models import BaseChatModel
from langchain_core.runnables.config import RunnableConfig
from langchain_core.tools import BaseTool

from langchain_mcp_adapters.client import MultiServerMCPClient

from llm_providers import LLMProvider, init_model_provider
from agent_tools import get_all_local_agent_tools


DEFAULT_CHAT_MODEL_TEMPERATURE = 0.


# --- The state class ---
class _WorkTrackerMCPAgentState(TypedDict):
    messages: Annotated[Sequence[BaseMessage], add_messages]
    loop_count: int


# --- The Agent class ---
class WorkTrackerMCPAgent:
    llm_provider: LLMProvider
    mcp_server_url: str
    chat_model: str
    chat_model_temperature: float

    _llm: BaseChatModel
    _llm_with_tools: Runnable[LanguageModelInput, AIMessage] | Any
    _graph: CompiledStateGraph[_WorkTrackerMCPAgentState, None, _WorkTrackerMCPAgentState, _WorkTrackerMCPAgentState]
    _memory_checkpointer: InMemorySaver

    # Initialization offloaded from the `__init__()` to `_init_mcp_and_graph()` called by the `create()` factory
    _mcp_client: MultiServerMCPClient

    _discovered_tools: list[BaseTool]
    _local_tools: list[BaseTool]  # The @tool decorator converts Callables into BaseTools
    _all_tools: list[BaseTool]
    _tools_registry: dict[str, BaseTool]

    _MAX_SIMULTANEOUS_TOOL_CALLS = 5
    _MAX_TOOL_CALLS_ROUNDS = 10  # Some of the user queries may involve many-round tool calls.

    def __init__(
        self, 
        llm_provider: LLMProvider, 
        chat_model: str,
        mcp_server_url: str,
        chat_model_temperature: float = DEFAULT_CHAT_MODEL_TEMPERATURE,
    ) -> None:
        """
        The constructor of the WorkTrackerMCPAgent class.
        NOT to be called directly, but only inside the `create()` factory method!
        """

        self.chat_model_temperature = chat_model_temperature
        self.chat_model = chat_model
        self.llm_provider = llm_provider
        self.mcp_server_url = mcp_server_url

        self._llm = init_model_provider(self.llm_provider, self.chat_model, self.chat_model_temperature)
        
        self._memory_checkpointer = InMemorySaver()

    @classmethod
    async def create(
        cls, 
        llm_provider: LLMProvider, 
        chat_model: str, 
        mcp_server_url: str,
        chat_model_temperature: float = DEFAULT_CHAT_MODEL_TEMPERATURE
    ) -> Self:
        """The async factory for the WorkTrackerMCPAgent."""
        instance = cls(llm_provider, chat_model, mcp_server_url, chat_model_temperature)
        await instance._init_mcp_and_graph()
        return instance

    async def run_single_query(self, user_message: str, thread_id: Optional[str] = None) -> str:
        """
        Process a single user query.

        Returns: The resulting Thread ID, which can be later used to resume the in-memory saved chat thread.
        """

        if thread_id is None:
            thread_id = str(uuid.uuid4())
        config = {"thread_id": thread_id}

        initial_state: _WorkTrackerMCPAgentState = {
            "messages": [HumanMessage(user_message)],
            "loop_count": 0
        }
        
        result = await self._graph.ainvoke(initial_state, config=RunnableConfig(configurable=config))

        print("\nAssistant: " + result["messages"][-1].content)

        return thread_id

    async def run_streaming_loop(self, cancellation_callback: Callable[[], bool], thread_id: Optional[str] = None) -> str:
        """
        Run a CLI streaming user chat loop.
        Exit from the loop by pressing Ctrl+C during the agent's thinking or answering,
        or by pressing Ctrl+D when waiting for a user input.

        Returns: The resulting Thread ID, which can be later used to resume the in-memory saved chat thread.
        """
        if thread_id is None:
            thread_id = str(uuid.uuid4())
        config = {"thread_id": thread_id}

        print("--- Session Started ---\n")
        print("In order to terminate the agent,"
              "\npress Ctrl+C when the agent is thinking or answering"
              "\nor Ctrl+D when waiting for the user input.\n")

        while not cancellation_callback():
            try:
                user_message = await asyncio.to_thread(input, "User:      ")
            except EOFError:
                break

            passed_state: _WorkTrackerMCPAgentState = {
                "messages": [HumanMessage(user_message)],
                "loop_count": 0
            }

            print("\nThinking...", end="\r")  # So that it can be rewritten

            prefix_printed = False
            async for part in self._graph.astream(
                    passed_state, config=RunnableConfig(configurable=config), stream_mode="messages", version="v2"):
                if part["type"] != "messages":
                    if cancellation_callback():
                        break
                    continue
                   
                msg, metadata = part["data"]

                # We target content chunks originating specifically from our 'agent' node
                if metadata.get("langgraph_node") == self._AGENT_NODE_NAME:
                    # Check if the chunk contains text content (ignores empty tool-call chunks)
                    if msg.content:
                        if not prefix_printed:
                            print("Assistant: ", end="", flush=True)
                            prefix_printed = True

                        print(msg.content, end="", flush=True)

                if cancellation_callback():
                    break
            
            if prefix_printed:
                print("\n\n", end="")

        return thread_id

    
    # --- MCP Discovery and Graph initialization ---
    async def _init_mcp_and_graph(self):
        """Asynchronously connects to the Work Tracker MCP server, discovers tools, and builds the graph."""
        print(f"Connecting to Work Tracker MCP server at '{self.mcp_server_url}'...")
        
        # Connect to the MCP server
        self._mcp_client = MultiServerMCPClient({
            "worktracker": {
                "transport": "streamable_http",
                "url": self.mcp_server_url
            }
        })
        
        # Automatically fetch all tools exposed by the MCP server (list_spaces, start_work_day, etc.)
        self._discovered_tools = await self._mcp_client.get_tools()

        # Combine local tools and remote MCP tools
        self._local_tools = get_all_local_agent_tools()
        self._all_tools = self._discovered_tools + self._local_tools
        
        # Build the tools registry and and bind them to the LLM
        self._tools_registry = {tool.name: tool for tool in self._all_tools}
        self._llm_with_tools = self._llm.bind_tools(list(self._tools_registry.values()))
        
        # Now that tools are bound, compile the graph
        self._graph = self._build_graph()
        
        print(f"Successfully connected! Discovered {len(self._discovered_tools)} MCP tools and adding {len(self._local_tools)} local tools.")


    # --- The nodes ---
    _AGENT_NODE_NAME = "agent"
    _TOOLS_NODE_NAME = "tools"

    async def _agent_node(self, state: _WorkTrackerMCPAgentState) -> dict[str, Any]:
        """Node: Invokes the LLM with the current conversation history, prepending the system message."""
        local_tools_names = [_tool.name for _tool in self._local_tools]
        local_tools_addend = f' along with a set of local tools (`{"`, `".join(local_tools_names)}`)' if len(local_tools_names) > 0 else ''

        system_message = textwrap.dedent(f"""
            You are a highly capable, integrated executive assistant connected directly to the user's Work Tracker application via the Model Context Protocol (MCP).
            Your primary role is to seamlessly manage their workspaces, tasks, work days, and time logging.

            ### Operational Guidelines:
            1. **Auto-Discovery:** 
                You have access to a suite of Work Tracker tools (e.g., `list_actionable_tasks`, `start_work_day`, `start_task_work`){local_tools_addend}. 
                Use them proactively to fulfill the user's intent.
            2. **State Dependencies:** 
                If the user asks to "start working on a task", you MUST check if there is an active work day using `get_current_work_status`.
                If there isn't one, invoke `start_work_day` before invoking `start_task_work`.
            3. **Data Retrieval:**
                If asked for a summary, utilize tools like `get_time_report` or `list_work_days` to provide accurate, up-to-date metrics. 
            4. **Destructive Actions:** 
                Proceed carefully with deletions (`delete_task`, `delete_work_day`). 
                If the user request is ambiguous, ask for confirmation before executing.
            5. **Efficiency:** 
                You may execute up to {self._MAX_SIMULTANEOUS_TOOL_CALLS} tool calls simultaneously and before giving an answer to the user, in only at most {self._MAX_TOOL_CALLS_ROUNDS} rounds. 
                Therefore if you were about to call tools in more than {self._MAX_TOOL_CALLS_ROUNDS}, answer to the user after at most {self._MAX_TOOL_CALLS_ROUNDS} tool call rounds and ask the user for a follow-up if they want to proceed.
                Do not invent task IDs; search for them if the user provides a name instead of an ID using `search_tasks`.

            Communicate concisely and professionally. You are assisting a power user who values speed and exactness.

            The current date and time is: {datetime.now().strftime('%A, %Y-%m-%d %H:%M:%S')}.
        """).strip()

        all_messages = [SystemMessage(system_message), *state["messages"]]

        # The LLM looks at the history and decides to either reply text or output a tool call.
        response = await self._llm_with_tools.ainvoke(all_messages)
        current_loops = state.get("loop_count", 0)
        return {"messages": [response], "loop_count": current_loops + 1}
        
    async def _execute_tools_node(self, state: _WorkTrackerMCPAgentState) -> dict[str, Any]:
        """Node: Manually loops through tool calls requested by the LLM."""
        last_message = state["messages"][-1]
        if not isinstance(last_message, AIMessage):
            return {}

        last_tool_calls_count = len(last_message.tool_calls)
        if last_tool_calls_count == 0:
            return {}  # Early exit

        # Programmatic Guardrail for max. parallelly queried items
        if last_tool_calls_count > self._MAX_SIMULTANEOUS_TOOL_CALLS:
            error_messages = []
            for tool_call in last_message.tool_calls:
                error_messages.append(
                    ToolMessage(
                        content=f"Error: Maximum of {self._MAX_SIMULTANEOUS_TOOL_CALLS} items allowed. You queried {last_tool_calls_count}. Please self-correct by prioritizing the most important tool calls, combining queries if possible, or executing them in sequential batches instead. Do not ask the user for help.", 
                        tool_call_id=tool_call["id"]
                    )
                )
            return {"messages": error_messages}
        
        # Setup lists to hold our async tasks and their corresponding IDs
        tasks = []
        tool_call_ids = []

        for tool_call in last_message.tool_calls:
            tool_func = self._tools_registry.get(tool_call["name"])
            if tool_func:
                # We call .ainvoke() universally. LangChain handles the sync-to-thread translation.
                tasks.append(tool_func.ainvoke(tool_call["args"]))
                tool_call_ids.append(tool_call["id"])
                
        # Execute all tools simultaneously, waiting for all to finish
        # return_exceptions=True prevents one failing tool from crashing the whole batch
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # Map the results back to ToolMessages
        tool_responses = []
        for result, call_id in zip(results, tool_call_ids):
            if isinstance(result, Exception):
                content = f"Tool execution failed: {str(result)}"
            else:
                content = str(result)
                
            tool_responses.append(ToolMessage(content=content, tool_call_id=call_id))
                
        return {"messages": tool_responses}
        
    def _route_or_break(self, state: _WorkTrackerMCPAgentState) -> str:
        """Conditional edge router: Custom router acting as a programmatic circuit breaker."""        
        # Circuit Breaker Triggered: Force exit if agent loops too many times
        if state.get("loop_count", 0) > self._MAX_TOOL_CALLS_ROUNDS:
            print(f"⚠️ Circuit breaker triggered! Agent exceeded {self._MAX_TOOL_CALLS_ROUNDS} loops.")
            return END
            
        # Standard operational routing path
        last_message = state["messages"][-1]
        if isinstance(last_message, AIMessage) and last_message.tool_calls:
            return self._TOOLS_NODE_NAME
            
        return END


    # --- The graph ---
    def _build_graph(self):
        """Builds the underlying LangGraph agent graph."""
        builder = StateGraph(_WorkTrackerMCPAgentState)

        # Nodes
        builder.add_node(self._AGENT_NODE_NAME, self._agent_node)
        builder.add_node(self._TOOLS_NODE_NAME, self._execute_tools_node)

        # Edges - a chain with the tool call loop
        builder.add_edge(START, self._AGENT_NODE_NAME)
        builder.add_conditional_edges(self._AGENT_NODE_NAME, self._route_or_break, {
            # Identity dictionary needed just to avoid reflection runtime errors
            self._TOOLS_NODE_NAME: self._TOOLS_NODE_NAME,
            END: END
        })
        builder.add_edge(self._TOOLS_NODE_NAME, self._AGENT_NODE_NAME)

        # Compile the graph
        return builder.compile(checkpointer=self._memory_checkpointer)
